### PR TITLE
StoragePoolType as class

### PR DESCRIPTION
**Pull request by sureshanaparti**
### Description

This PR moves the StoragePoolType from an Enum to a class that is largely compatible with Enum. This allows for storage pool plugins to be added dynamically without needing to hard code an enum value into the core code. We are trying to be Enum compatible to avoid having to refactor much of the code where these are used.

To do this, on the KVM agent side we needed to move from annotation based StoragePoolTypes on the StorageAdaptors to having an adaptor method report the StoragePoolType. This was because annotations aren't dynamic enough to be able to use StoragePoolTypes that aren't static.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by kiranchavala** [comment]
@blueorangutan test matrix

**Comment by DaanHoogland** [comment]
good functionality @sureshanaparti , I haven't looked at the code yet, but good to see movements like this, making ACS more configurable/extensible(/expandable?)

**Comment by sureshanaparti** [comment]
@blueorangutan package

File: framework/db/src/main/java/com/cloud/utils/db/GenericDaoBase.java [hunk 1]
```diff
@@ -650,7 +655,7 @@ protected void setField(Object entity, Field field, ResultSet rs, int index) thr
             } else if (type == byte[].class) {
                 field.set(entity, rs.getBytes(index));
             } else {
-                field.set(entity, rs.getObject(index));
+                field.set(entity, _conversionSupport.convertToEntityAttribute(field, rs.getObject(index)));
```
Comment (DaanHoogland) [inline framework/db/src/main/java/com/cloud/utils/db/GenericDaoBase.java:658]: is this unconditionally? below a check for the annotation is done (see line 1603)

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/storage/LibvirtStorageAdaptor.java [hunk 2]
```diff
@@ -775,23 +781,22 @@ public KVMPhysicalDisk createPhysicalDisk(String name, KVMStoragePool pool,
...
-                        return createPhysicalDiskByQemuImg(name, pool, format, provisioningType, size, passphrase);
-                    case DIR:
-                    case TAR:
-                        return createPhysicalDiskByLibVirt(name, pool, format, provisioningType, size);
-                    default:
-                        throw new CloudRuntimeException("Unexpected disk format is specified.");
-                }
-            default:
-                return createPhysicalDiskByLibVirt(name, pool, format, provisioningType, size);
+        StoragePoolType poolType = pool.getType();
```
Comment (DaanHoogland) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/storage/LibvirtStorageAdaptor.java:784]: can we use `switch (poolType /*.toString*/)` here? jabva supports switching on strings, so no reason to do this refactor.

**Review by DaanHoogland — APPROVED** [review]
looks good. one question and one not so pressing remark.

**Review by kiranchavala — APPROVED** [review]
LGTM, Tested by adding various types of primary storage and performed basic operations

**Comment by rohityadavcloud** [comment]
@blueorangutan test matrix

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/storage/LibvirtStorageAdaptor.java (hunk 2 above)
Comment (sureshanaparti) [reply plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/storage/LibvirtStorageAdaptor.java:784]: tried switch/case on poolType (string), noticed some errors with it. keeping the new/refactored changes.

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test matrix

**Review by JoaoJandre — APPROVED** [review]
CLGTM, didn't test it

**Review by kiranchavala — APPROVED** [review]
LGTM, tested manually was able to add cluster of type  "DatastoreCluster"

**Review by harikrishna-patnala — APPROVED** [review]
LGTM, tested with some storage related operations

**Pull request merged**